### PR TITLE
More HTML and JavaScript. Functional and Computer Programming.

### PR DESCRIPTION
**Pull request by JStonham**
Please can you take a look at my 28.02.19 and 07.03.19 notes? :)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2217,4 +2217,219 @@ A Japanese word meaning "form", relating to karate. Used commonly in programming
...
+
+HTML (Hypertext Markup Language) forms the structure of web pages.
+CSS (Cascading Style Sheets) forms the style.
+JavaScript provides the dynamic and interactive parts.
+
+### JavaScript
+
+JavaScript started as a browser language, but now it is used for a variety of things including: Slack, Facebook Messenger, Web Servers and Applications.
+
+The JavaScript execution engine is Node. No HTML is required, but CSS is.
```
Comment (kinbiko) [inline README.md]: The code we wrote was in Node, and we didn't write any HTML nor CSS :)

```suggestion
When using JavaScript outside of the browser you generally use Node.js.
```

File: README.md [hunk 2]
```diff
@@ -2217,4 +2217,219 @@ A Japanese word meaning "form", relating to karate. Used commonly in programming
...
+
+As JavaScript does not have types, it does not have type safety. Microsoft addressed the issues of this by creating TypeScript, which includes all the capabilities of JavaScript and additionally includes types.
```
Comment (kinbiko) [inline README.md]: Javascript does have types (in fact you list the important ones below ;)). Only very few, really odd programming languages (e.g. bash) are untyped.

```suggestion
As JavaScript does not have a compile-time type system, and as such it does not have type safety. Microsoft addressed the issues of this by creating TypeScript, which includes all the capabilities of JavaScript and additionally includes types.
```

File: README.md [hunk 3]
```diff
@@ -2217,4 +2217,219 @@ A Japanese word meaning "form", relating to karate. Used commonly in programming
...
+- Boolean
+- Functions
+- null
+- undefined
+- Array
+- Object
+
+### HTML
+
+HTML is not a programming language. It does have classes and IDs for referencing components that programmers wish to reassign. IDs must be assigned only one value where as classes can have multiple values (e.g. if programmers want many aspects to have the same formatting).
```
Comment (kinbiko) [inline README.md]:
```suggestion
HTML is not a programming language. It does have classes and IDs for referencing components that programmers wish to reassign. IDs must be assigned only once whereas classes can be assigned to multiple elements (e.g. if programmers want many aspects to have the same formatting).
```

File: README.md [hunk 4]
```diff
@@ -2217,4 +2217,219 @@ A Japanese word meaning "form", relating to karate. Used commonly in programming
...
+
+HTML and pictures have been combined to create Scalar Vector Graphics (SVG).
+
+The basic format for HTML is:
```
Comment (kinbiko) [inline README.md]:
```suggestion
In very broad terms, the basic format for HTML is:
```

File: README.md [hunk 5]
```diff
@@ -2217,4 +2217,219 @@ A Japanese word meaning "form", relating to karate. Used commonly in programming
...
+
+#### HTML vs Markdown 
+
+| Markdown | HTML |
+| -------- | ---- |
+|  `#`  | `<h1>` |
+|  `-`  | `<ul> <li> text goes here </li> </ul>` |
+|  `1.` | `<ol> <li> text goes here </li> </ol>` |
+| backtick | `<pre>` |
+| ` ```java | `<pre><code></code></pre>` |
```
Comment (kinbiko) [inline README.md]: I don't think this renders quite right. Maybe

```suggestion
| ` ```java ` | `<pre><code></code></pre>` |
```

File: README.md [hunk 6]
```diff
@@ -2217,4 +2217,219 @@ A Japanese word meaning "form", relating to karate. Used commonly in programming
...
+| `**` | `<b>` |
+
+## Functional Programming with JavaScript
+
+Functional Programming is the combination of declarative coding (telling code what you want it to do) and immutable (unchangeable) data.
```
Comment (kinbiko) [inline README.md]:
```suggestion
Functional Programming is the combination of declarative coding (telling code what you want it to do (as opposed to *how* to do it)) and immutable (unchangeable) data.
```

File: README.md [hunk 7]
```diff
@@ -2217,4 +2217,219 @@ A Japanese word meaning "form", relating to karate. Used commonly in programming
...
+
+JavaScript allows programmers to put the output of one function into another function.
```
Comment (kinbiko) [inline README.md]:
```suggestion
JavaScript allows programmers pass functions into other functions.
```

File: README.md [hunk 8]
```diff
@@ -2217,4 +2217,219 @@ A Japanese word meaning "form", relating to karate. Used commonly in programming
...
+
+An example of a variable within a function:
+
+```javascript
+let j = function(myVar) {
+    myVar("hello")
+}
+```
+
+In JavaScript, for loops become `forEach` and include a variable name.
```
Comment (kinbiko) [inline README.md]:
```suggestion
In JavaScript, although you can use the classic for-loop with an index variable, most people use `forEach`:
```

File: README.md [hunk 9]
```diff
@@ -2217,4 +2217,219 @@ A Japanese word meaning "form", relating to karate. Used commonly in programming
...
+The following example assigns 5 numbers to an array, then filters that array to return only even numbers. Then the even numbers are assigned to `foo` and printed. Lastly, the entire array is printed.
+
+```javascript
+let array = [1,2,4,6,7]
+
+array.filter(function(n){ return n % 2 === 0}) // classic function
+    .map(n => "foo: " + n) // arrow function
+    .forEach(n => console.log(n))
+
+console.log(array);
```
Comment (kinbiko) [inline README.md]: I don't think this line adds much
```suggestion
```

File: README.md [hunk 10]
```diff
@@ -2217,4 +2217,219 @@ A Japanese word meaning "form", relating to karate. Used commonly in programming
...
+```
+
+It's important to note that you should never change the nature of an array once it's been defined (e.g. once `array =` you should never write `array =` again in that code).
```
Comment (kinbiko) [inline README.md]:
```suggestion
In strict functional programming, you should never change the nature of an array once it's been defined (e.g. once `array =` you should never write `array =` again in that code).
The benefit of doing so is that all the state of your application can never change from underneath you and introduce confusing bugs.
```

File: README.md [hunk 11]
```diff
@@ -2217,4 +2217,219 @@ A Japanese word meaning "form", relating to karate. Used commonly in programming
...
+        return "Buzz"
+    }
+    if (i % 3 === 0) { 
+        return "Fizz"
+    }
+    return i
+}).forEach(i => console.log(i))
+```
+
+## Computer Programming
```
Comment (kinbiko) [inline README.md]:
```suggestion
## Program Complexity
```

File: README.md [hunk 12]
```diff
@@ -2217,4 +2217,219 @@ A Japanese word meaning "form", relating to karate. Used commonly in programming
...
+
+Order = the rate of growth of a function.
+
+| Order | Time |
+| ----- | ---- |
+| O(1) | fast |
+| O(log n) |  `|` |
+| O(nlog n) | `|`  |
+| O(n) |  `|` |
+| O(n<sup>2</sup>) | `|`  |
```
Comment (kinbiko) [inline README.md]: nice! You're using HTML!

File: README.md [hunk 13]
```diff
@@ -2217,4 +2217,219 @@ A Japanese word meaning "form", relating to karate. Used commonly in programming
...
+```
+
+## Computer Programming
+
+Order = the rate of growth of a function.
+
+| Order | Time |
+| ----- | ---- |
+| O(1) | fast |
+| O(log n) |  `|` |
```
Comment (kinbiko) [inline README.md]: Did you forget to fill these out?

File: README.md [hunk 14]
```diff
@@ -2217,4 +2217,219 @@ A Japanese word meaning "form", relating to karate. Used commonly in programming
...
+| O(nlog n) | `|`  |
+| O(n) |  `|` |
+| O(n<sup>2</sup>) | `|`  |
+| O(2<sup>n</sup>) | `v`  |
+| O(n!) | slow |
+
+`O(n)` is used for searching unsorted lists.
```
Comment (kinbiko) [inline README.md]: Saying that it's "used for" is not quite right IMO. It makes it sound like we can choose any of these always, which isn't the case. It's more that `O(n)` is the best we can do in this scenario.

**Review by kinbiko — APPROVED** [review]
Good notes as always! Looks pretty good to me. Nothing major so feel free to address without assigning the review to me again.

File: README.md (hunk 13 above)
Comment (JStonham) [reply README.md]: I was trying to do an arrow pointing down the column from fast to slow, but maybe I'll do some words like 'very fast', 'fast' etc instead!

**Pull request merged**